compensations/src/main/java/org/jboss/narayana/compensations/impl/CompensationContextLifetimeManager.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.narayana.compensations.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Class used to make sure that CompensationContext is closed once all participants have completed.
+ *
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+public final class CompensationContextLifetimeManager {
+
+    private static CompensationContextLifetimeManager INSTANCE = new CompensationContextLifetimeManager();
+
+    private final Map<Object, AtomicInteger> counters;
+
+    public static CompensationContextLifetimeManager getInstance() {
+        return INSTANCE;
+    }
+
+    private CompensationContextLifetimeManager() {
+        this.counters = new HashMap<>();
+    }
+
+    /**
+     * Notify the manager that new participant was created.
+     *
+     * @param currentTX
+     */
+    public synchronized void participantCreated(final Object currentTX) {
+        AtomicInteger counter = counters.get(currentTX);
+
+        if (counter == null) {
+            counter = new AtomicInteger(1);
```

Comment:
Ok, it must be something subtle because it seems the get on line 53 would retrieve the counter so it should already be in the counters map and hence shouldn't need the put on line 61?